collector/meminfo_linux.go

```diff
@@ -17,59 +17,189 @@
 package collector
 
 import (
-	"bufio"
 	"fmt"
-	"io"
-	"os"
-	"regexp"
-	"strconv"
-	"strings"
-)
 
-var (
-	reParens = regexp.MustCompile(`\((.*)\)`)
+	"github.com/go-kit/log"
+	"github.com/prometheus/procfs"
 )
 
-func (c *meminfoCollector) getMemInfo() (map[string]float64, error) {
-	file, err := os.Open(procFilePath("meminfo"))
+type meminfoCollector struct {
+	fs     procfs.FS
+	logger log.Logger
+}
+
+// NewMeminfoCollector returns a new Collector exposing memory stats.
+func NewMeminfoCollector(logger log.Logger) (Collector, error) {
+	fs, err := procfs.NewFS(*procPath)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("failed to open procfs: %w", err)
 	}
-	defer file.Close()
 
-	return parseMemInfo(file)
+	return &meminfoCollector{
+		logger: logger,
+		fs:     fs,
+	}, nil
 }
 
-func parseMemInfo(r io.Reader) (map[string]float64, error) {
-	var (
-		memInfo = map[string]float64{}
-		scanner = bufio.NewScanner(r)
-	)
+func (c *meminfoCollector) getMemInfo() (map[string]float64, error) {
+	meminfo, err := c.fs.Meminfo()
+	if err != nil {
+		return nil, fmt.Errorf("Failed to get memory info: %s", err)
+	}
+
+	metrics := make(map[string]float64)
 
-	for scanner.Scan() {
-		line := scanner.Text()
-		parts := strings.Fields(line)
-		// Workaround for empty lines occasionally occur in CentOS 6.2 kernel 3.10.90.
-		if len(parts) == 0 {
-			continue
-		}
-		fv, err := strconv.ParseFloat(parts[1], 64)
-		if err != nil {
-			return nil, fmt.Errorf("invalid value in meminfo: %w", err)
-		}
-		key := parts[0][:len(parts[0])-1] // remove trailing : from key
-		// Active(anon) -> Active_anon
-		key = reParens.ReplaceAllString(key, "_${1}")
-		switch len(parts) {
-		case 2: // no unit
-		case 3: // has unit, we presume kB
-			fv *= 1024
-			key = key + "_bytes"
-		default:
-			return nil, fmt.Errorf("invalid line in meminfo: %s", line)
-		}
-		memInfo[key] = fv
+	if meminfo.ActiveBytes != nil {
```

Comment:
I'd convert this to a loop, something like:

```golang
for k, v := range map[string]*int {..} {
  if v == nil {
    continue
  }
  metrics[k] = float64(*v)
```

Reply:
I think we wanted to switch to an explicit mapping.

Reply:
Agreed, I think we need an explicit mapping here, especially since the field names in the Meminfo struct don't 100% align with the existing metric names, ex:

```	if meminfo.ActiveAnonBytes != nil {
		metrics["Active_anon_bytes"] = float64(*meminfo.ActiveAnonBytes)```